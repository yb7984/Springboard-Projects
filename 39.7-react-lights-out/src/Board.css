.Board {
  margin: 0 auto;
  background-color: lightseagreen;
}

.Board-won{
  margin: 0 auto;
  text-align: center;
  font-size: 3em;
  background-color: gold;
  color: red;
  font-weight: 500;
  height: 200px;
  align-items: center;
  display: flex;
  justify-content: center;
}