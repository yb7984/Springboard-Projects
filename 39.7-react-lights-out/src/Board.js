import React, { useState } from "react";
import Cell from "./Cell";
import "./Board.css";

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/

function Board({ nrows, ncols, chanceLightStartsOn }) {
  const [board, setBoard] = useState(createBoard());

  /** create a board nrows high/ncols wide, each cell randomly lit or unlit */
  function createBoard() {
    let initialBoard = [];
    // create array-of-arrays of true/false values
    console.log(Math.random());

    for (let y = 0; y < nrows; y++) {
      const row = [];
      for (let x = 0; x < ncols; x++) {
        row.push(Math.random() < chanceLightStartsOn ? true : false);
      }
      initialBoard.push(row);
    }

    return initialBoard;
  }

  function hasWon() {
    //  check the board in state to determine whether the player has won.
    return !board.some(row => row.some(cell => cell === true));
  }

  function flipCellsAround(coord) {
    setBoard(oldBoard => {
      const [y, x] = coord.split("-").map(Number);

      const flipCell = (y, x, boardCopy) => {
        // if this coord is actually on board, flip it
        if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
          boardCopy[y][x] = !boardCopy[y][x];
        }
      };

      // Make a (deep) copy of the oldBoard
      const boardCopy = oldBoard.map(row => [...row]);

      // in the copy, flip this cell and the cells around it
      for (let i = y - 1; i <= y + 1; i++) {
        for (let j = x - 1; j <= x + 1; j++) {
          flipCell(i, j, boardCopy);
        }
      }

      return boardCopy;
    });
  }

  // if the game is won, just show a winning msg & render nothing else
  if (hasWon()) {
    return (<div className="Board-won">You Won</div>)
  }

  // make table board
  return (<table className="Board" data-testid="Board">
    <tbody>
      {
        board.map((row, y) => {
          return (
            <tr key={y}>
              {
                row.map((isLit, x) => {
                  return <Cell cellId={`${y}-${x}`} key={`${y}-${x}`} flipCellsAroundMe={flipCellsAround} isLit={isLit} />

                })
              }
            </tr>
          )
        })
      }</tbody>
  </table>)
}

Board.defaultProps = {
  nrows: 5,
  ncols: 5,
  chanceLightStartsOn: 0.5
}

export default Board;
