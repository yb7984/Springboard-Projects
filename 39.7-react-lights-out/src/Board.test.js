import React from "react";
import { render, fireEvent, asFragment } from "@testing-library/react";
import Board from "./Board";

//make thae random just out put 0.2 and 0.8 one after another
let current = 0.8;

Math.random = () => {
    if (current === 0.8) {
        current = 0.8;
        return 0.2;
    }
    current = 0.2;
    return 0.2;
}

it("renders without crashing", function () {
    render(<Board />);
});


it("matches snapshot", function () {
    current = 0.8;
    const { asFragment } = render(<Board />);
    expect(asFragment()).toMatchSnapshot();
});

it("matches snapshot , all lights off, show the win message", function () {
    const { asFragment, queryByText, queryByTestId } = render(<Board chanceLightStartsOn={0} />);
    expect(asFragment()).toMatchSnapshot();

    //show the win message
    expect(queryByText("You Won", { exact: false })).toBeInTheDocument();
    //not showing the board
    expect(queryByTestId("Board")).not.toBeInTheDocument();
});



it("matches snapshot , all lights on, show the board", function () {
    const { asFragment, queryByText, queryByTestId } = render(<Board chanceLightStartsOn={1} />);
    expect(asFragment()).toMatchSnapshot();

    //show the win message
    expect(queryByText("You Won", { exact: false })).not.toBeInTheDocument();
    //not showing the board
    expect(queryByTestId("Board")).toBeInTheDocument();

    //showing all light on
    for (let y = 0; y < 5; y++) {
        for (let x = 0; x < 5; x++) {
            expect(queryByTestId(`${y}-${x}`)).toHaveClass("Cell-lit");
        }
    }
});


it("works when click a cell", function () {
    current = 0.8;
    const { asFragment, queryByText, queryByTestId } = render(<Board chanceLightStartsOn={1} />);

    //all lights on
    let clickCell = queryByTestId("2-2");
    for (let y = 1; y <= 3; y++) {
        for (let x = 1; x <= 3; x++) {
            expect(queryByTestId(`${y}-${x}`)).toHaveClass("Cell-lit");
        }
    }

    fireEvent.click(clickCell);

    //all light around it turn off, others stay on
    for (let y = 0; y < 5; y++) {
        for (let x = 0; x < 5; x++) {
            if (y >= 1 && y <= 3 && x >= 1 && x <= 3) {
                expect(queryByTestId(`${y}-${x}`)).not.toHaveClass("Cell-lit");
            } else {
                expect(queryByTestId(`${y}-${x}`)).toHaveClass("Cell-lit");
            }
        }
    }

});
