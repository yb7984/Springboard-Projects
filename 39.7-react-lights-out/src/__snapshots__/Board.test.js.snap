// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot , all lights off, show the win message 1`] = `
<DocumentFragment>
  <div
    class="Board-won"
  >
    You Won
  </div>
</DocumentFragment>
`;

exports[`matches snapshot , all lights on, show the board 1`] = `
<DocumentFragment>
  <table
    class="Board"
    data-testid="Board"
  >
    <tbody>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="0-0"
        />
        <td
          class="Cell Cell-lit"
          data-testid="0-1"
        />
        <td
          class="Cell Cell-lit"
          data-testid="0-2"
        />
        <td
          class="Cell Cell-lit"
          data-testid="0-3"
        />
        <td
          class="Cell Cell-lit"
          data-testid="0-4"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="1-0"
        />
        <td
          class="Cell Cell-lit"
          data-testid="1-1"
        />
        <td
          class="Cell Cell-lit"
          data-testid="1-2"
        />
        <td
          class="Cell Cell-lit"
          data-testid="1-3"
        />
        <td
          class="Cell Cell-lit"
          data-testid="1-4"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="2-0"
        />
        <td
          class="Cell Cell-lit"
          data-testid="2-1"
        />
        <td
          class="Cell Cell-lit"
          data-testid="2-2"
        />
        <td
          class="Cell Cell-lit"
          data-testid="2-3"
        />
        <td
          class="Cell Cell-lit"
          data-testid="2-4"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="3-0"
        />
        <td
          class="Cell Cell-lit"
          data-testid="3-1"
        />
        <td
          class="Cell Cell-lit"
          data-testid="3-2"
        />
        <td
          class="Cell Cell-lit"
          data-testid="3-3"
        />
        <td
          class="Cell Cell-lit"
          data-testid="3-4"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="4-0"
        />
        <td
          class="Cell Cell-lit"
          data-testid="4-1"
        />
        <td
          class="Cell Cell-lit"
          data-testid="4-2"
        />
        <td
          class="Cell Cell-lit"
          data-testid="4-3"
        />
        <td
          class="Cell Cell-lit"
          data-testid="4-4"
        />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <table
    class="Board"
    data-testid="Board"
  >
    <tbody>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="0-0"
        />
        <td
          class="Cell Cell-lit"
          data-testid="0-1"
        />
        <td
          class="Cell Cell-lit"
          data-testid="0-2"
        />
        <td
          class="Cell Cell-lit"
          data-testid="0-3"
        />
        <td
          class="Cell Cell-lit"
          data-testid="0-4"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="1-0"
        />
        <td
          class="Cell Cell-lit"
          data-testid="1-1"
        />
        <td
          class="Cell Cell-lit"
          data-testid="1-2"
        />
        <td
          class="Cell Cell-lit"
          data-testid="1-3"
        />
        <td
          class="Cell Cell-lit"
          data-testid="1-4"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="2-0"
        />
        <td
          class="Cell Cell-lit"
          data-testid="2-1"
        />
        <td
          class="Cell Cell-lit"
          data-testid="2-2"
        />
        <td
          class="Cell Cell-lit"
          data-testid="2-3"
        />
        <td
          class="Cell Cell-lit"
          data-testid="2-4"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="3-0"
        />
        <td
          class="Cell Cell-lit"
          data-testid="3-1"
        />
        <td
          class="Cell Cell-lit"
          data-testid="3-2"
        />
        <td
          class="Cell Cell-lit"
          data-testid="3-3"
        />
        <td
          class="Cell Cell-lit"
          data-testid="3-4"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="4-0"
        />
        <td
          class="Cell Cell-lit"
          data-testid="4-1"
        />
        <td
          class="Cell Cell-lit"
          data-testid="4-2"
        />
        <td
          class="Cell Cell-lit"
          data-testid="4-3"
        />
        <td
          class="Cell Cell-lit"
          data-testid="4-4"
        />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
